net/pfSense-pkg-WireGuard/files/usr/local/pkg/wireguard/wg.inc

```diff
@@ -0,0 +1,416 @@
+		exec_wg_quick_action($tunnel, "up", $verbose);
+
+	}
+
+	// Now let's always soft configure the interface
+	wg_syncconf($tunnel);
+
+	// Add the interface to the interface group (i.e. WireGuard)
+	exec("{$wgg['ifconfig']} {$tunnel['name']} group {$wgg['ifgroupentry']['ifname']}" . verbose_pipe($verbose));
+
+	// Get the tunnel interface friendly name if assigned
+	$if_fname = convert_real_interface_to_friendly_interface_name($tunnel['name']);
+
+	if(isset($if_fname)) {
+
+		interface_reconfigure($if_fname);
+
+		system_routing_configure($if_fname);
+	
+	}
+
+}
+
+
+// Remove WireGuard tunnel
+function wg_destroy_if($tunnel, $verbose = false) {
+
+	// Execute the wg-quick(8) way of tearing down an interface
+	exec_wg_quick_action($tunnel, "down", $verbose);
+
+	// Need to transform out of tunnel structure because pfSense_interface_destroy expects an ifname
+	if (is_array($tunnel) && isset($tunnel['name'])) {
+
+		$tunnel = $tunnel['name'];
+
+	}
+
+	// Now execute the pfSense way of tearing down an interface
+	pfSense_interface_destroy($tunnel);
+
+}
+
+function wg_delete_tunnel($tunidx) {
+	global $config, $wgg;
+
+	wg_globals();
+
+	if ($wgg['tunnels'][$tunidx]) {
+
+		unset($ifname);
+
+		unset($conf_path);
+
+		if (isset($wgg['tunnels'][$tunidx]['name'])) {
+
+			$ifname = $wgg['tunnels'][$tunidx]['name'];
+
+		}
+		
+		if (isset($ifname)) {
+
+			$conf_path = $wgg['conf_path'] . "/" . $ifname . ".conf";
+
+		}
+
+		// Delete the tunnel configuration entry
+		unset($wgg['tunnels'][$tunidx]);
+
+		write_config("[WireGuard] tunnel {$index} updated.");
+
+		// Delete the wg?.conf file
+
+		if (isset($conf_path) && is_file($conf_path)) {
+
+			unlink($conf_path);
+
+		}
+
+		// Destroy the deleted tunnel
+		wg_destroy_if($ifname);
+	}
+}
+
+// Write new tunnel values to the configuration system
+function wg_do_post($post) {
+	global $config, $wgg;
+
+	wg_globals();
+
+	$input_errors = array();
+
+	$index = $post['index'];
+
+	$pconfig = $wgg['tunnels'][$index];
+
+	if (empty($pconfig['name'])) {
+
+		$pconfig['name'] = next_wg_if();
+
+	}
+
+	if (empty($pconfig['mtu'])) {
+
+		$pconfig['mtu'] = $wgg['default_mtu'];
+
+	}
+
+	$pconfig['enabled'] = empty($post['enabled']) ? 'no' : 'yes';
+
+	$pconfig['descr'] = $post['descr'];
+
+	// Tunnel section
+	$pconfig['interface']['address'] = $post['address'];
+	$pconfig['interface']['listenport'] = $post['listenport'];
+	$pconfig['interface']['privatekey'] = $post['privatekey'];
+	$pconfig['interface']['publickey'] = genPubKey($post['privatekey']);
+
+	$pconfig['interface']['mtu'] = $post['mtu'];
+
+	if (isset($post['endpoint0'])) {
+		// Peers section
+		$pconfig['peers'] = array();
+		$pconfig['peers']['wgpeer'] = array();
+
+		$idx = 0;
+
+		for (;;) {
+			if (!$post['publickeyp'.$idx] &&
+			    !$post['endpoint'.$idx] &&
+			    !$post['allowedips'.$idx] &&
+			    !$post['descr'.$idx] &&
+			    !$post['persistentkeepalive'.$idx]) {
+				break;
+			}

```

Comment:
This code would be more clean using a while statement instead of infinite for with a conditional break

```php
		while ($post['publickeyp'.$idx] &&
		    $post['endpoint'.$idx] &&
		    $post['allowedips'.$idx] &&
		    $post['descr'.$idx] &&
		    $post['persistentkeepalive'.$idx]) {
```

Reply:
I actually copy-pasted this in my revision (my bad) and of course it broke haha, I took some time to refactor this by setting the input field names to `<input name="endpoint[]"` in vpn_wg_edit.php. This removes the need to guess/check looping and we can use a for loop instead.

```
	if (!empty($post['endpoint'])) {

		// Peers section
		$pconfig['peers'] = array();

		$pconfig['peers']['wgpeer'] = array();

		for ($idx = 0; $idx <= count($post['endpoint']); $idx++) {

			$peer = array();
			$peer['publickey'] = $post['publickeyp'][$idx];
			$peer['endpoint'] = $post['endpoint'][$idx];
			$peer['allowedips'] = $post['allowedips'][$idx];
			$peer['descr'] = $post['descp'][$idx];
			$peer['persistentkeepalive'] = $post['persistentkeepalive'][$idx];
			$peer['presharedkey'] = $post['presharedkey'][$idx];
			$peer['port'] = $post['port'][$idx];
			$peer['peerwgaddr'] = $post['peerwgaddr'][$idx];

			$pconfig['peers']['wgpeer'][] = $peer;

		}

	} else {
		unset($pconfig['peers']);
	}
```